ion_schema_2_0/constraints/regex.isl

```diff
@@ -0,0 +1,1091 @@
+  regex: "[bcd]",
+}
+$test::{
+  type: regex_character_class,
+  should_accept_as_valid: [
+    "b",
+    "c",
+    "d",
+  ],
+  should_reject_as_invalid: [
+    "a",
+    "e",
+  ]
+}
+
+type::{
+  name: regex_character_range_negated,
+  regex: "[^b-g]",
+}
+$test::{
+  type: regex_character_range_negated,
+  should_accept_as_valid: [
+    "a",
+    "i",
+  ],
+  should_reject_as_invalid: [
+    "b",
+    "c",
+    "d",
+    "e",
+    "f",
+    "g",
+  ]
+}
+
+type::{
+  name: regex_character_range,
+  regex: "[b-g]",
+}
+$test::{
+  type: regex_character_range,
+  should_accept_as_valid: [
+    "b",
+    "c",
+    "d",
+    "e",
+    "f",
+    "g",
+  ],
+  should_reject_as_invalid: [
+    "a",
+    "i",
+  ]
+}
+
+type::{
+  name: regex_complex_nested_quantified_alternation,
+  regex: "^(ab|(cd){2,3}|ef){2}$",
+}
+$test::{
+  type: regex_complex_nested_quantified_alternation,
+  should_accept_as_valid: [
+    abab,
+    abcdcd,
+    abcdcdcd,
+    abef,
+    cdcdab,
+    cdcdcdab,
+    cdcdcdcd,
+    cdcdcdcdcd,
+    cdcdcdcdcdcd,
+    cdcdcdef,
+    cdcdef,
+    efab,
+    efcdcd,
+    efcdcdcd,
+    efef,
+  ],
+  should_reject_as_invalid: [
+    "",
+    ab,
+    abcdcdcdcd,
+    cd,
+    cdcd,
+    cdcdcd,
+    cdcdcdcdcdcdcd,
+    cdcdcdcdef,
+    ef,
+  ]
+}
+
+type::{
+  name: regex_escape_characters,
+  // There is an space at the end of this regex because the Ion plugin doesn't handle escaped backsalshes at the end
+  // of a string correctly. See https://github.com/amzn/ion-intellij-plugin/issues/40.
+  regex: "\\.\\^\\$\\|\\?\\*\\+\\[\\]\\(\\)\\{\\}\\\\ ",
+}
+$test::{
+  type: regex_escape_characters,
+  should_accept_as_valid: [
+    '.^$|?*+[](){}\\ ',
+    ".^$|?*+[](){}\\ ",
+  ],
+  should_reject_as_invalid: [
+    null,
+    null.null,
+    null.string,
+    null.symbol,
+    null.bool,
+    'hello world',
+    "hello world",
+  ]
+}
+
+type::{
+  name: regex_escape_double_quote,
+  regex: "\"free\"",
+}
+$test::{
+  type: regex_escape_double_quote,
+  should_accept_as_valid: [
+    '"free"',
+    "\"free\"",
+    'Totally "free" food!',
+    "Totally \"free\" food!",
+  ],
+  should_reject_as_invalid: [
+    null,
+    null.null,
+    null.string,
+    null.symbol,
+    null.bool,
+    '',
+    "",
+    'free',
+    "free",
+  ]
+}
+
+type::{
+  name: regex_escape_forwardslash,
+  regex: "a/b",
+}
+$test::{
+  type: regex_escape_forwardslash,
+  should_accept_as_valid: [
+    'a/b',
+    "a/b",
+  ],
+  should_reject_as_invalid: [
+    null,
+    null.null,
+    null.string,
+    null.symbol,
+    null.bool,
+    '',
+    "",
+    ab,
+    "ab",
+  ]
+}
+
+type::{
+  name: regex_escape_single_quote,
+  regex: "it's",
+}
+$test::{
+  type: regex_escape_single_quote,
+  should_accept_as_valid: [
+    'it\'s',
+    "it's",
+    'I can\'t believe it\'s not butter!',
+    "I can't believe it's not butter!",
+  ],
+  should_reject_as_invalid: [
+    null,
+    null.null,
+    null.string,
+    null.symbol,
+    null.bool,
```

Comment:
We repeat a lot that regular expressions don't match various flavors of `null` or empty strings. I think I'd rather see a couple of the most permissive regular expressions not matching `null` flavors and then omit them from future test cases. It would be very surprising for `"it's"` to change the relationship of regex to `null`.

Reply:
Sure, I can do that.